include/bbp/sonata/reports.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+/**
+ * Provides a bridge between c-based programs in order to have access to the c++-based objects
+ */
+#ifndef __cplusplus
```

Comment:
To add here, somewhere else I saw a return-an-int-to-indicate-error. This would probably improved by basing the C++ code on exceptions and only returning error-code stuff in the C interface.

Reply:
> This is expected to be a C++ library; if it needs to be wrapped in a c interface, that can go in the bindings/ directory under bindings/c.

Agree!

> The main interface, IMO, should be exposed as a C++ one

Can we push this in separate PR/issue?

Reply:
Why push it to a separate issue/PR? The public interface and its stability is probably the most important thing in a library, so I'd prefer to get it right when adding new code to a library.

Reply:
> The public interface and its stability is probably the most important thing 

Agree with this!

> Why push it to a separate issue/PR? 

I think this is more coming from confusion about below:

> The main interface, IMO, should be exposed as a C++ one

 Currently we are consuming this all through C like API. If we are talking about exposing C++ (polished) public API that can be consumed by others then I am saying that we should work on that separately. At the end that will be something like [class SonataReport](https://github.com/BlueBrain/libsonata/blob/0e64b09b38bad3040d9fa1e41b10dd221786fafd/src/reports/library/sonatareport.h) but jus want to bring the time constraints we have to implement writers (this PR) and implement readers.

Reply:
> Currently we are consuming this all through C like API (which I considered as public one).

I think I see it differently then - this is a C++ library that can have bindings; its interface is therefore C++.  However, one can wrap the *public* part of the C++ API in other languages, like we have done for python.  This wrapping should only use the C++ API functions, and no internal ones, unless there is a strong case of an exception.  If things aren't done this way, we're making what should be private implementation details public, with the consequences this brings (think Hyrum's Law.)

> bring the time constraints we have to implement writers (this PR) and implement readers.

Creating the readers is can be a separate PR to this, and is what is *for sure* needed in libsonata for the MOOC.  Making the reading PR isn't contingent on this PR, IMO, and it would actually be quite helpful to know whether writing needs to be in this library - it gives data about the question of what the overlap/benefits are, compared to the costs.